added a python script file for cotrip translator and changed requirment.txt file

abinashkonersman opened the pull request.

payneBrandon committed:
Merge pull request #19 from CDOT-CV/dev

Dev

abinashkonersman committed:
added a python script file for cotrip translator and changed requirement.txt file

abinashkonersman committed:
Update README.md

updated documentation for cotrip translator

payneBrandon commented on README.md:
```diff
@@ -34,7 +34,7 @@ pip install -r requirements.txt
-### Execution
+### Execution for iCone translator
 
 #### Step 1: Run the translator script (from Work_Zone/translator/source code/Folder)
```
The trailing `/Folder` isn't needed here.

payneBrandon commented on README.md:
```diff
@@ -46,6 +46,18 @@ Example usage:
 
+### Execution for COtrip translator
+
+#### Step 1: Run the translator script (from Work_Zone/translator/source code/Folder)
```
Same here, trailing `/Folder` is incorrect.

payneBrandon commented on README.md:
````diff
@@ -46,6 +46,18 @@ Example usage:
+```
+Example usage:
+```
+python cotrip_translator.py -i '../sample files/icone data/incidents_extended.xml' 
````
Running this command produces the following error:
![image](https://user-images.githubusercontent.com/2932159/113169850-3a8fec80-9203-11eb-8302-0b637454cc37.png)

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+
```
Can we organize this file a bit better, it's difficult to follow. For instance we have the usage and what would appear to be the entry point in the middle of the file. Let's move that to the top or bottom to call it out from the rest of the code. Also, there's an if statement floating in the file and not in a method, let's pull that into a "main" method or something similar.

PTRFRLL replied:
Agreed, we should have our "main" method near the top and any functions that the main functions uses, should follow. That way the file can be read "top down"

abinashkonersman replied:
Should be fixed now

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+
+
+if inputfile:
+    # Added encoding argument because of weird character at start of alerts.xml file
```
Encoding argument? Where is that?

abinashkonersman replied:
deleted the comment that was a comment from icone_translator

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+if inputfile:
+    # Added encoding argument because of weird character at start of alerts.xml file
+
+    icone_obj = parse_xml(inputfile)
```
This isn't an icone object, it's an XML object parsed from from cotrip at this point. Let's rename accordingly.

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    with open(inputfile, encoding='utf-8-sig') as ficone:
+        # Read
+        xml_string = ficone.read()
+        icone_obj = xmltodict.parse(xml_string)
```
Same up here, this isn't an icone object.

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+      return False
+    return True
+
+def validate_write(wzdx_obj, outputfile, location_schema):
```
This function makes it sound like you are validating the write of the file. Can we break it into two functions to have each method with a single responsibility? IE one to validate, and another to write the file?

abinashkonersman replied:
i deleted the method and added the code in main method

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+
```
Several of these methods seem to be the same as in `icone_translator.py`. Let's create a shared library that we can pull in to both of those scripts and reuse the code.

abinashkonersman replied:
i created a shared library in source_code directory.

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+
+    event_status = "active"
+    if datetime.now(timezone.utc) < start_time:
+        event_status = "planned"  # if < 2 to 3 weeks make it pending instead of planned
```
I don't see where we set this as "pending"?

abinashkonersman replied:
i updated the function for event_status = "pending"

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+        return icone_obj
+
+def validate_wzdx(wzdx_obj, wzdx_schema):
+    #wzdx_schema = json.loads(open(location_schema).read())
```
Can we remove?

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    #### Consider whether this id needs to be hardcoded or generated
+    info['feed_info_id'] = "104d7746-688c-44ed-b195-2ee948bf9dfa"
+
+    #### This information is required, might want to hardcode
```
Let's update this comment. We are already hardcoding the values.

abinashkonersman replied:
should be fixed now

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    info = {}
+
+    #### Consider whether this id needs to be hardcoded or generated
+    info['feed_info_id'] = "104d7746-688c-44ed-b195-2ee948bf9dfa"
```
This should probably come from an environment setting, or config file. Currently this is the same id as in the icone_translator file, which means we are telling our applications the data is from the same location which is inaccurate.

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+        logging.warning('invalid type')
+        return False
+    
+    #### Consider whether this id needs to be hardcoded or generated
```
This comment and the one below just seem to be repeats of when we are creating the info object. Let's either update to make sense in this context, or remove.

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    for field in required_fields:
+        if not field:
+            return False
+            logging.warning( 'Not all required fields are present') 
```
This logging statement is never hit, we return before it is called.

abinashkonersman replied:
i moved the logging statement to its proper position

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    'Left Turn Lane Closure', 'Center Lanes Closed', 'Right Turn Lane Closure', 'Alternating Single Lane Closures','Intermittent Lane Closure'  ]
+    all_lanes_open = ['No Closure', -'Intermittent Traffic Stops', 'Shoulder Closure', 'Right Shoulder Closed', 'Off-Ramp Closure', 'On-Ramp Closure',  ]
+    alternating_one_way = ['Single Lane Traffic', 'Alternating Traffic' ]
+    unknown = []
```
Variable is never used, let's remove.

abinashkonersman replied:
removed the variable

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+# function to calculate vehicle impact
+def get_vehicle_impact(closure_type):
+
+        #Closure type value : [-'Shoulder Closure', -'Single Lane Closure', -'No Closure', -'Right Shoulder Closed', -'Alternating Traffic', '-Off-Ramp Closure',
```
All these values are laid out below, can we remove this comment?

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+
+
+# Parse COtrip alert to WZDx
+def parse_alert(alert, callback_function=None):
```
Can we cleanup the comments in this method? There seems to be quite a few that are not needed.

abinashkonersman replied:
cleaned up the comment

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    # start_date
+    properties['start_date'] = reformat_datetime(alert['alert:ReportedTime'])
+
+    # end_date
```
We are using two different access methods on the alert object here, can we stick to one for consistency?

PTRFRLL replied:
This seems to be the case throughout the file, let's try to standardize on an access method

jacob6838 replied:
Is this referring to alert['event'] vs alert.get('event')?

payneBrandon replied:
Yes

payneBrandon replied:
@abinashkonersman any update on this one?

abinashkonersman replied:
updated all to .get()

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+
+    # vehicle impact
+
+    properties['vehicle_impact'] = 'unknown' #get_vehicle_impact(alert['alert:ClosureType'])
```
The `get_vehicle_impact` method exists up above, is there a reason we aren't calling it?

abinashkonersman replied:
The data we intended to use to calculate the get vehicle impact is not what we thought and at this point we can not calculate it. Thats why we set it 'unknown'. possibly in future we will use this function.

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    properties['lanes'] = []
+
+    # road_name
+    properties['road_number'] = ""
```
Are we not able to parse this from the `alert:RoadName` property?

jacob6838 replied:
It is usually possible to parse the road_number from the RoadName, but not always. This could definitely be added, not sure what the logic would be, maybe just if the road name has a number in it?

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    properties['types_of_work'] = []
+
+    # reduced speed limit
+    properties['reduced_speed_limit'] = 25
```
This doesn't seem accurate. I see various speed limits in the example data.

abinashkonersman replied:
should be fixed

payneBrandon replied:
It looks like the `reduced_speed_limit` setting was just removed...does the cotrip not specify if there is a reduced limit? The example data seemed to have speed limit values in it.

payneBrandon replied:
@abinashkonersman any word on this? It appears to be the same in the `icone_translator.py` where we are had coding the value to 25 for some reason.

abinashkonersman replied:
should be fixed

payneBrandon replied:
Again, the `reduced_speed_limit` was just removed, but our example data does have speed limit values in it that we should be taking into account.

abinashkonersman replied:
I don't see any reduced_speed_limit in our example data(cotrip_1.json). can you please specify which data i should consider for reduced_speed_limit?

payneBrandon replied:
It looks like we had reduced speed data in the original [xml file](https://raw.githubusercontent.com/CDOT-CV/Work_Zone/a172aab5ba67ec23ea810bd4ca006e81f55e50ad/translator/sample%20files/cotrip_data/cotrip_data.xml). Do we not have similar data to this in the newer JSON files?

abinashkonersman replied:
you are right and  i found some in  cotrip xml file and working on it

abinashkonersman replied:
created breakout functions for reduced speed limit  which looks for the speed limit in description and writes it to wzdx message and added required unit test for those function.

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+    properties['reduced_speed_limit'] = 25
+
+    # workers present
+    properties['workers_present'] = False
```
Can we leave this blank if we don't know whether workers are present or not? I do see descriptions saying flaggers present in the sample data.

abinashkonersman replied:
should be  fixed

payneBrandon replied:
This seems to be the same in the `icone_translator.py` where we are hard coding to False. If we don't know the value, let's not set it.

abinashkonersman replied:
should be fixed now in icone_translator.py

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+            return False
+            
+    try:
+        dateutil.parser.parse(alert['alert:ReportedTime'])
```
We already grabbed the ReportedTime and stored it in a variable above, let's reuse.

abinashkonersman replied:
I used the variable starttime there

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+            
+    try:
+        dateutil.parser.parse(alert['alert:ReportedTime'])
+        if alert.get('alert:ExpectedEndTime'):
```
Same here, we know that ExpectedEndTime exists from above, and can reuse that `endtime` variable here.

abinashkonersman replied:
I changed that to end time

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+#### This function may fail if some optional fields are not present (lanes, types_of_work, relationship, ...)
+def add_ids(message, add_ids):
+    if add_ids:
+        feed_info_id = message['road_event_feed_info']['feed_info_id']
```
Variable is never used.

abinashkonersman replied:
i removed the unused variable

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,421 @@
+        road_event_length = len(message['features'])
+        road_event_ids = []
+        for i in range(road_event_length):
+            road_event_ids.append(str(uuid.uuid4()))
```
Are we just using this block to generate a random road event id? We could skip this and just add a random one below.

jacob6838 replied:
This exists for the case where there is more than one feature per work zone. In that case, the properties.relationship object will contain all of the road_event_ids for the related features. It is possible we will soon implement something like this, for each of the work_update/schedule objects in the cotrip message, so this may be of use soon. Currently it is not adding anything though

abinashkonersman committed:
added sample files

abinashkonersman committed:
removed the  trailing /Folder from README.md

abinashkonersman committed:
Renamed icone_obj to cotrip_obj in code

abinashkonersman committed:
Removed the commented line from line number 392

abinashkonersman committed:
Removed the commented line

abinashkonersman committed:
removed some unnecessary comment

abinashkonersman committed:
renamed from icone to cotrip

abinashkonersman committed:
changed the relative path

abinashkonersman committed:
updated the execution instruction in README.md

PTRFRLL commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,409 @@
+
+
+def parse_arguments(argv):
+    inputfile = ''
```
since `inputFile` is required for the program to do anything, perhaps we can print the help message if the input flag (`-i`) is not provided

abinashkonersman replied:
i updated the code in help message

payneBrandon replied:
This is now handled for the `cotrip_translator.py` file, but not with `icone_translator.py`.

abinashkonersman replied:
i did it in icone_translator.py too

payneBrandon commented:
Can we add an entry to the .gitignore for the cotrip output file?

PTRFRLL commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,409 @@
+    wzd['road_event_feed_info']['feed_info_id'] = info['feed_info_id']
+    wzd['road_event_feed_info']['update_date'] = datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ")
+    wzd['road_event_feed_info']['publisher'] = 'CDOT'
+    wzd['road_event_feed_info']['contact_name'] = 'Abinash Konersman'
```
These hardcoded values (`contact_name`, `contact_email`, etc) are used later in the file as well, can we make them a variable so we only need to change them once?

abinashkonersman replied:
should be fixed now

payneBrandon replied:
These values are referenced in only one location now which is good, but Peter is right - we should pull them out to an environment variable or script argument so we can easily change things like contact email in the future.

abinashkonersman replied:
set those as environment variable and set the environment variable in env_var.sh file in translator/source_code so before we run our program we have to run the .sh script file in our local computer then run the program.

abinashkonersman committed:
Added the deduplication script python program

abinashkonersman committed:
updated gitignore for cotrip wzdx  message geojson file

abinashkonersman committed:
Update cotrip_wzdx_translated_output_message.geojson

abinashkonersman committed:
added main method and organized the code

abinashkonersman committed:
removed unused variable unknown = []

abinashkonersman committed:
Added a shared library for icone and cotrip translator file

abinashkonersman committed:
added the condition for event_status = pending in get_event_status method

abinashkonersman committed:
removed some unnecessary comments

abinashkonersman committed:
removed some of the unnecessary comments and unused variables

abinashkonersman committed:
commit

abinashkonersman committed:
Added Shared_library and modified cotrip translator

abinashkonersman committed:
added some unit test

abinashkonersman committed:
added uniit test for cotrip and shared_library file

abinashkonersman committed:
updated environment variable to default

abinashkonersman committed:
updated organization name to CDOT

abinashkonersman committed:
Merge pull request #24 from CDOT-CV/CV-92-2

Cv 92 2

payneBrandon commented:
@abinashkonersman It's been a while since we've opened this and there doesn't seem to be a description of what this PR entails. Can you add a description to this so I know what we're looking at?

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,296 @@
+
+import xmltodict
```
Quite a few of these imports are no longer used, can we clean up?

abinashkonersman replied:
should be fixed

payneBrandon replied:
Not fixed, there's still numerous imports that are not used in here.

abinashkonersman replied:
i removed it ..hopefully it should be fixed now

payneBrandon replied:
Still numerous unused imports in this `cotrip_translator.py` file as well as other files. Let's try to clean that up.

abinashkonersman replied:
removed unused imports

payneBrandon replied:
We still have several import statements throughout the code base that are unused.

abinashkonersman replied:
i hoped i deleted all

payneBrandon commented on translator/source_code/translator_shared_library.py:
```diff
@@ -0,0 +1,95 @@
+import xmltodict
```
Quite a few of these imports are not used, let's clean it up a bit.

abinashkonersman replied:
cleaned it up.should be fixed

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,296 @@
+from collections import OrderedDict
+import re
+import dateutil.parser
+import translator_shared_library
```
Because of the way the local files are imported, it makes navigation around the code base difficult. Can we import them with the full namespace so we can include navigation?  IE
```
from translator.source_code import translator_shared_library
```

This should apply to unit tests and other modules as well.

abinashkonersman replied:
should be fixed now

abinashkonersman committed:
removed some of the unnecessary imports

PTRFRLL commented on README.md:
````diff
@@ -46,6 +46,18 @@ Example usage:
+```
+Example usage:
+```
+python cotrip_translator.py -i '../sample files/cotrip_data/cotrip_data.xml' 
````
Is `cotrip_data.xml` supposed to be in this repo? I don't see it.

abinashkonersman replied:
should be fixed now

PTRFRLL commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+    inputfile, outputfile = translator_shared_library.parse_arguments(sys.argv[1:], default_output_file_name = 'cotrip_wzdx_translated_output_message.geojson')
+    if inputfile:
+
+        cotrip_obj = json.loads(open(inputfile).read())
```
Looks like we're expecting the input file to be json. Can we document that assumption in the `help_string` and add a check if they supply the wrong filetype? (eg. if i set inputfile as XML, the program crashes)

abinashkonersman replied:
should be fixed

PTRFRLL replied:
I can still supply an XML file and it crashes the program. At the very least, we should specify that the input file NEEDS to be in JSON format. 


should be fixed

abinashkonersman replied:
should be fixed now

payneBrandon replied:
I'm still not seeing this requirement in the help message, nor mentioned in the readme.

abinashkonersman replied:
i mentioned that in README.md 

Execution for COtrip translator
Step 1: Run the translator script (from Work_Zone) and Please specify a valid Json file as an input file !
python -m translator.source_code.cotrip_translator -i inputfile.json -o outputfile.geojson
Example usage:

python -m translator.source_code.cotrip_translator -i 'translator/sample files/cotrip_data/cotrip_1.json'

PTRFRLL commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+    coordinates = []
+    for i in polyline:
+        coords = i.split(' ')
+        if len(coords) >= 2 and re.match('^-?([0-9]*[.])?[0-9]+$', coords[0]) and re.match('^-?([0-9]*[.])?[0-9]+$', coords[1]):
```
Perhaps add a comment detailing what the Regex is looking for.

abinashkonersman replied:
I added a comment

PTRFRLL commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+
+    # type_of_work
+    # maintenance, minor-road-defect-repair, roadside-work, overhead-work, below-road-work, barrier-work, surface-work, painting, roadway-relocation, roadway-creation
+    properties['types_of_work'] = []
```
Are we not adding `types_of_work` or `restrictions`? Using `cotrip_1.json` as an example, I see "roadway-relocation" and "Flagger Operations" but they are excluded from the output file.

PTRFRLL replied:
Were these properties purposely excluded?

payneBrandon replied:
@abinashkonersman If we've got those values in the input file, we'll want them as part of our output.

abinashkonersman replied:
should be fixed

PTRFRLL commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+
+# Add ids to message
+#### This function may fail if some optional fields are not present (lanes, types_of_work, relationship, ...)
+def add_ids(message, add_ids):
```
Do we need the `add_ids` argument? Seems like we always call this function with it set to True

abinashkonersman replied:
i deleted the add_ids argument

PTRFRLL commented on translator/source_code/translator_shared_library.py:
```diff
@@ -0,0 +1,95 @@
+
+help_string = """ 
+
+Usage: python icone_translator.py [arguments]
```
If I run the cotrip_translator without args I get a message detailing how to run the icone_translator:
```
Usage: python icone_translator.py [arguments]
```
We should probably convert to a function and allow passing in the program name

abinashkonersman replied:
i modified the usage message

PTRFRLL commented on translator/source_code/deduplicate.py:
```diff
@@ -0,0 +1,67 @@
+    lat1, lon1 = origin  # lat/lon of origin
+    lat2, lon2 = destination  # lat/lon of dest
+
+    radius = 6371.0*1000  # meters
```
We might consider updating the variable names to be more descriptive [[1](https://moderatemisbehaviour.github.io/clean-code-smells-and-heuristics/#g19-use-explanatory-variables)], e.g. `radiusInMeters`, it helps avoid mental mapping

abinashkonersman replied:
i deleted that file since we are not using that deduplication anymore

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+        location_schema = '../sample files/validation_schema/wzdx_v3.0_feed.json'
+        wzdx_schema = json.loads(open(location_schema).read())
+        if not translator_shared_library.validate_wzdx(wzdx_obj, wzdx_schema):
+            print('validation error more message are printed above')
```
If we fail on validation, we should return at this point. Currently, even on validation fail we are opening the output file and writing to it.

abinashkonersman replied:
i changed the print message.should be fixed

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+    wzd = {}
+    wzd['road_event_feed_info'] = {}
+    # hardcode
+    wzd['road_event_feed_info']['feed_info_id'] = info['feed_info_id']
```
There seems to be a fair bit of duplicated assignments between here and the `icone_translator.py` file when creating the wzdx message. Let's abstract what we can to the shared library. For instance this entire hard-coded section of the `road_event_feed_info` assignment can be pushed to the library and likely much more.

payneBrandon commented on translator/source_code/icone_translator.py:
```diff
@@ -21,8 +22,8 @@ def wzdx_creator(messages, info=None, unsupported_message_callback=None):
-        info=initialize_info()
-    if not validate_info(info):
+        info = translator_shared_library.initialize_info()
+    if not translator_shared_library.validate_info(info):
```
We're checking for nulls and validating up here which is great. Let's move the check for message incidents below (~line 64) up to here as well. That will prevent all this code being called if we have bad data in the first place.

abinashkonersman replied:
should be fixed

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+            callback_function(alert)
+        return None
+    event = alert.get('event', {})
+    feature = {}
```
This is overridden below, do we really need it here?

abinashkonersman replied:
i removed it .it should be fixed

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+    feature['type'] = "Feature"
+    properties = {}
+
+    # road_event_id
```
If this is handled in the add_ids call, do we really need to set it to empty here?

jacob6838 replied:
This preserves the order of the message. It is kind of stupid but it makes the message match the order of the documentation

payneBrandon replied:
If that's what we're using it for, how about adding a new function `initialize_properties` that creates a properties object with all the properties set to blank. Then in here, we just set the properties we know about? That way all the properties are set, and its a bit more clear as to what we're doing.

payneBrandon replied:
@abinashkonersman @jacob6838 any progress on this one?

abinashkonersman replied:
added a function initialize_features_properties and it should be fixed now

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+
+    # data_source_id
+    #### Leave this empty, it will be populated by add_ids
+    properties['data_source_id'] = ''
```
Same here, do we really need to set?

abinashkonersman replied:
should be fixed

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+    #### Leave this empty, it will be populated by add_ids
+    properties['data_source_id'] = ''
+
+    # start_date
```
Quite a few of these comments still aren't getting us anything. We can already see that this is setting the 'start_date' value, and don't need the comment. Same for most of the below comments.

jacob6838 replied:
I like these comments because they break up the code, and make it obvious which code is for what field, especially when a field uses a multi-line expression

abinashkonersman replied:
@payneBrandon  do you still want to remove these comments?

payneBrandon commented on translator/source_code/cotrip_wzdx_translated_output_message.geojson:
```diff
@@ -0,0 +1,100 @@
+{
```
Do we need this file? Can we remove?

abinashkonersman replied:
i removed that file

payneBrandon commented on translator/source_code/deduplicate.py:
```diff
@@ -0,0 +1,67 @@
+import json
```
I don't see any unit tests for this file?

payneBrandon replied:
Also, what are we de-duplicating? Are there duplicate messages we need to get rid of? I was under the impression that we would be merging our icone data into the cotrip data which would require finding the matching sets and combining data, not removing anything.

jacob6838 replied:
That is exactly what was intended, de-duplicating is inaccurate. This is being put on hold because we do not have any duplicate iCone/cotrip messages, so this can be deleted

payneBrandon commented on translator/sample files/Output Message/cotrip_wzdx_translated_output_message.geojson:
```diff
@@ -0,0 +1,100 @@
+{
```
Do we need this file? Can we remove?

abinashkonersman replied:
i removed that file

payneBrandon commented on translator/sample files/cotrip_data/cdot-rtdh-prod_cotrip-events-standard-oim-wzdx-integration_20210316-232928-207252.json:
```diff
@@ -0,0 +1,36 @@
+{
```
There's several of these files that seem to be named with a date, is this something we're using? If so, can we add description to the README and if not, can we remove?

abinashkonersman replied:
removed all those unnecessary files

payneBrandon commented on translator/sample files/cotrip_data/cotrip_1.json:
```diff
@@ -0,0 +1,108 @@
+{
```
Is this file used?

abinashkonersman replied:
we are using it to run the program

payneBrandon replied:
It's never mentioned anywhere. If we have multiple test input files, let's update the documentation to reflect this.

abinashkonersman replied:
i used that file name in documentation..in fact i deleted the cotrip_2.json from that directory.


### Execution for COtrip translator

#### Step 1: Run the translator script (from Work_Zone/translator/source code)

```
python cotrip_translator.py -i inputfile.json -o outputfile.geojson
```
Example usage:
```
python cotrip_translator.py -i '../sample files/cotrip_data/cotrip_1.json'

payneBrandon replied:
Ah! I was thinking of `cotrip_2.json` below. I do see the reference to `cotrip_1.json` though.

payneBrandon commented on translator/sample files/cotrip_data/cotrip_2.json:
```diff
@@ -0,0 +1,36 @@
+{
```
Is this file used?

abinashkonersman replied:
i removed that file.

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,294 @@
+import uuid
+import random
+import string
+import sys, getopt
```
Several of these imports are unused, can we clean up?

abinashkonersman replied:
i removed the unused imports and that extension worked very well.

payneBrandon replied:
Still quite a few unused imports.

abinashkonersman replied:
deleted those imports

payneBrandon replied:
`cotrip_translator.py` still has a couple unused imports:
![image](https://user-images.githubusercontent.com/2932159/118304491-f5b6d100-b4a3-11eb-940c-9eeb437b7e30.png)

Here it appears that "ge" and "validate" are not used.

abinashkonersman replied:
deleted those.should be fixed now

payneBrandon commented on test/test_cotrip.py:
```diff
@@ -0,0 +1,561 @@
+
+
+
+def test_wzdx_creator_empty_icone_object() :
```
We're not testing icone, let's rename accordingly.

abinashkonersman replied:
should be fixed

payneBrandon commented:
Some of these files have mixed indentation and spacing, are you using a formatter? For VSCode I recommend the following extensions to help:
- https://marketplace.visualstudio.com/items?itemName=ms-python.python
- https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance

payneBrandon commented on test/test_cotrip.py:
```diff
@@ -0,0 +1,561 @@
+  assert test_wzdx == None
+
+def test_wzdx_creator_invalid_info_object() :
+  cotrip_obj = cotrip_obj = {
```
I don't think this double assignment is necessary.

abinashkonersman replied:
removed the double assignment.

payneBrandon commented on test/test_cotrip.py:
```diff
@@ -0,0 +1,561 @@
+  test_valid_output = 'invalid output'
+  assert cotrip_translator.validate_alert(test_valid_output) == False
+
+def test_validate_incident_no_data():
```
Should be named validate alert, not incident.

abinashkonersman replied:
should be fixed now

payneBrandon commented on test/test_cotrip.py:
```diff
@@ -0,0 +1,561 @@
+}
+  assert cotrip_translator.validate_alert(test_valid_output) == True
+
+def test_validate_incident_missing_required_field_description():
```
Should be named validate alert, not incident.

abinashkonersman replied:
changed it to test_validate_alert_missing_required_field_description

payneBrandon commented on test/test_cotrip.py:
```diff
@@ -0,0 +1,561 @@
+    test_time = "16093s98000z"
+    actual_time = cotrip_translator.reformat_datetime(test_time)
+    expected_time = ''
+    assert actual_time == expected_time
```
Missing tests for add_ids method

abinashkonersman replied:
i added unit test for add_ids method

payneBrandon commented on test/test_shared_library.py:
```diff
@@ -0,0 +1,98 @@
+from translator.source_code import translator_shared_library
+import json
+import pytest
+import unittest
```
Unused, can we remove?

abinashkonersman replied:
removed import unittest

payneBrandon commented on test/test_shared_library.py:
```diff
@@ -0,0 +1,98 @@
+  assert test_validate_info == False
+
+#--------------------------------------------------------------------------------unit test for parse_xml function--------------------------------------------------------------------------------
+def test_parse_xml():
```
Good success case, can we add a check for failed cases?

abinashkonersman replied:
added some unit test

payneBrandon replied:
Looks like this added a case where the passed in value is empty. Can we add one more to test with a non-existent file name since we're checking for that in the code? IE something like 'non-file.txt'.

abinashkonersman replied:
added another test case for non-file.txt

payneBrandon commented on translator/source_code/translator_shared_library.py:
```diff
@@ -0,0 +1,95 @@
+def parse_xml(inputfile):
+    with open(inputfile, encoding='utf-8-sig') as ficone:
+        xml_string = ficone.read()
+        icone_obj = xmltodict.parse(xml_string)
```
Since this is for more than icone now, let's rename those variables accordingly.

abinashkonersman replied:
should be fixed now

payneBrandon commented on test/test_shared_library.py:
```diff
@@ -0,0 +1,98 @@
+    validate_write=translator_shared_library.validate_wzdx(valid_wzdx_data,test_schema)
+    assert validate_write == True
+
+def test_validate_write_invvalid_wzdx_data():
```
Spelling issue - only one 'v' in invalid. Also, we're testing the `validate_wzdx` method not `validate_write`.

abinashkonersman replied:
should be fixed now

abinashkonersman committed:
removed unused 'import unittest'

payneBrandon commented on test/test_shared_library.py:
```diff
@@ -0,0 +1,98 @@
+        translator_shared_library.parse_arguments(test_argv)
+
+#--------------------------------------------------------------------------------unit test for validate_write function--------------------------------------------------------------------------------
+def test_validate_wzdx_valid_wzdx_data():
```
Can we add a few more cases here, for instance what happens if the schema is empty, or the json file is?

abinashkonersman replied:
i added some edge cases

payneBrandon commented on test/test_shared_library.py:
```diff
@@ -0,0 +1,98 @@
+from translator.source_code import translator_shared_library
```
Missing a test for the `initialize_info` method.

abinashkonersman replied:
added unit test for initialize_info

abinashkonersman committed:
corrected the spelling of invalid and changed the function name to test_validate_wzdx_valid_wzdx_data

payneBrandon commented:
Is `deduplicated_wzdx_field` file supposed to have a file extension? Is it even needed? I see it in the .gitigore file but if we want it gone, we'll need to check in a delete of the file.

abinashkonersman committed:
changed method name to test_validate_incident_missing_required_field_description in test_cotrip.py

abinashkonersman committed:
changed name to validate_alert

abinashkonersman committed:
removed the double assignment

abinashkonersman committed:
renamed it to test_wzdx_creator_empty_cotrip_object

abinashkonersman committed:
renamed test_wzdx_creator_empty_cotrip_object

abinashkonersman committed:
deleted deduplicated.py file

abinashkonersman committed:
Update translator_shared_library.py

payneBrandon commented on test/test_cotrip.py:
```diff
@@ -0,0 +1,559 @@
+
+  
+def test_wzdx_creator_invalid_incidents_no_description() :
+  cotrip_obj = cotrip_obj = {
```
Duplicate assignment.

abinashkonersman replied:
should be fixed

payneBrandon replied:
Not fixed, still duplicated saying `cotrip_obj = cotrip_obj = ...`

abinashkonersman replied:
i removed it from the test file

abinashkonersman committed:
changed gitignore

abinashkonersman committed:
removed some unnecessary imports imports

abinashkonersman committed:
Delete cotrip_2.json

abinashkonersman committed:
added unit test for add ids method

abinashkonersman committed:
added unit test for parse_xml function

abinashkonersman committed:
added some unit test

PTRFRLL commented on .gitignore:
```diff
@@ -4,4 +4,7 @@
 icone_to_wzdx_test.geojson
-wzdx_translated_output_message.geojson
\ No newline at end of file
+icone_wzdx_translated_output_message.geojson
```
For future reference, you can use wildcards `*` to simplfy this to:
```
*.geojson
```

abinashkonersman replied:
should be fixed

abinashkonersman committed:
added mapping docs

abinashkonersman committed:
updated gitignore

abinashkonersman committed:
added script file for combine wzdx

abinashkonersman committed:
updated combine_wzdx.py

abinashkonersman committed:
deleted combine_wzdx.py file which was not required for this story

abinashkonersman committed:
added *.pyc to gitignore file

abinashkonersman committed:
deleted the unnecessary file

payneBrandon commented on translator/source_code/translator_shared_library.py:
```diff
@@ -0,0 +1,171 @@
+
+def initialize_info():
+    info = {}
+    info['feed_info_id'] = "104d7746-688c-44ed-b195-2ee948bf9dfa"
```
We're still hard coding this feed info id to the same value for both cotrip and icone data. This would then make our output data appear to be coming from the same location, which doesn't seem correct.

abinashkonersman replied:
should be fixed

payneBrandon commented on test/test_shared_library.py:
```diff
@@ -0,0 +1,255 @@
+    assert validate_write == True
+
+
+def test_validate_wzdx_invalid_wzdx_data():
```
What's invalid here? Can we be more descriptive in the test name so we know what it is we are testing?

abinashkonersman replied:
i renamed the unit test name

payneBrandon reviewed (changes requested):
Added several comments that need addressing before merge.

abinashkonersman commented:
> Added several comments that need addressing before merge.

Working on those comments.Thanks

abinashkonersman committed:
updated gitignore file

abinashkonersman committed:
changed the access methods to .get() to to one for consistency

abinashkonersman committed:
removed unused imports

abinashkonersman committed:
removed cotrip_obj = cotrip_obj

abinashkonersman committed:
specified what type of file we should give as an input (.json)file

abinashkonersman committed:
added a breakout function i.e get_types_of_work

abinashkonersman committed:
added different feed info id for icone and cotrip

abinashkonersman committed:
renamed the unit test name to more specific name

abinashkonersman committed:
added some unit test

abinashkonersman committed:
updated and added some unit test

PTRFRLL commented:
Latest changes look good to me. I like the breakout of the `get_types_of_work`. Can you publish this PR so I can add my approval?

abinashkonersman commented:
> Is `deduplicated_wzdx_field` file supposed to have a file extension? Is it even needed? I see it in the .gitigore file but if we want it gone, we'll need to check in a delete of the file.

that file is no longer exist

payneBrandon commented on translator/source_code/cotrip_translator.py:
```diff
@@ -0,0 +1,327 @@
+    if not sub_type or type(sub_type) != str:
+        return []
+    sub_type_split = sub_type.split(':')
+    if len(sub_type_split) < 2:
```
Is it not possible to have a single work type?

payneBrandon replied:
Ah I see, the json file has them split like this. Makes sense now, closing.

abinashkonersman committed:
removed some unused imports

abinashkonersman committed:
removed unused import patch

abinashkonersman committed:
fixed missing positional parameter

abinashkonersman committed:
updated the README.md

abinashkonersman committed:
Update README.md

abinashkonersman committed:
changed the print the help message if the input flag (-i) for icone_translator.py

abinashkonersman committed:
added contact_name,contact_emailenvirontment variable

abinashkonersman committed:
added environment variable

abinashkonersman committed:
updated README.md for environment variable set up and added some unit test

abinashkonersman committed:
updated README.md

abinashkonersman committed:
updated README.md

abinashkonersman committed:
removed on unused import

abinashkonersman commented:
Hi @payneBrandon  i think i addressed all those comments you made. can please review , so that we can merge that into dev and we can start working for the transition of WZDX 3.1

abinashkonersman committed:
added the updated wzdx mapping docs

payneBrandon commented:
> Hi @payneBrandon i think i addressed all those comments you made. can please review , so that we can merge that into dev and we can start working for the transition of WZDX 3.1

@abinashkonersman unit tests are failing: 
![image](https://user-images.githubusercontent.com/2932159/118895468-753d0980-b8c3-11eb-8dc7-65d30dbd5a5a.png)

It looks like we aren't mocking the newly created environment variables.

payneBrandon commented on README.md:
```diff
@@ -32,7 +32,25 @@ This code requires Python 3.6 or a higher version. If you haven’t already, dow
+
+
+Example usage:
+for mac computer run the following script to iniliatize the environment variable:
```
`initialize` rather than `iniliatize`.

abinashkonersman replied:
made it initialize

abinashkonersman committed:
mocked some unit test function for environment variable

abinashkonersman committed:
updated unit test

abinashkonersman commented:
@payneBrandon  mocked the environment variable in our test function.The PR should be ready to review now. Thanks.

payneBrandon commented on README.md:
```diff
@@ -32,18 +32,48 @@ This code requires Python 3.6 or a higher version. If you haven’t already, dow
 
+#### Environment variable
 
+Please set up the following environmet variable for your local computer before running the script.
```
`environment` rather than `environmet`.

abinashkonersman replied:
corrected it to environment

payneBrandon commented on README.md:
```diff
@@ -32,18 +32,48 @@ This code requires Python 3.6 or a higher version. If you haven’t already, dow
+| :---        |    :----:   |          ---: |
+| contact_name     | Abinash Konersman       |  name of environment variable contact_name containing name of the cotact person  |
+| contact_email   | abinash.konersman@state.co.us       | name of contact email containing email  of contact person |
+| issuing_organization   | CDOT       | name of the oraganization     |
```
This should be `organization` rather than `oraganization`.

abinashkonersman replied:
should be fixed

abinashkonersman commented:
> Can we add an entry to the .gitignore for the cotrip output file?

should be fixed

abinashkonersman committed:
corrected the spelling mistakes in README.md
